mlflow/server/js/src/components/CompareRunScatter.js

```diff
@@ -0,0 +1,222 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import { getParams, getRunInfo } from '../reducers/Reducers';
+import { connect } from 'react-redux';
+import './CompareRunView.css';
+import { RunInfo } from '../sdk/MlflowMessages';
+import Utils from '../utils/Utils';
+import { getLatestMetrics } from '../reducers/MetricReducer';
+import {ScatterChart, Scatter, XAxis, YAxis, CartesianGrid, Tooltip, ResponsiveContainer, Label} from 'recharts';
+import './CompareRunScatter.css';
+
+class CompareRunScatter extends Component {
+  static propTypes = {
+    runInfos: PropTypes.arrayOf(RunInfo).isRequired,
+    metricLists: PropTypes.arrayOf(Array).isRequired,
+    paramLists: PropTypes.arrayOf(Array).isRequired,
+  };
+
+  constructor(props) {
+    super(props);
+
+    this.metricKeys = CompareRunScatter.getKeys(this.props.metricLists);
+    this.paramKeys = CompareRunScatter.getKeys(this.props.paramLists);
+
+    if(this.paramKeys.length + this.metricKeys.length < 2)
+      this.state = {disabled: true};
+    else
+      this.state = {
+        disabled: false,
+        x: this.paramKeys.length > 0 ?
+          {
+            key: this.paramKeys[0],
+            isMetric: false
+          } : {
+            key: this.metricKeys[1],
+            isMetric: true
+          },
+        y: this.metricKeys.length > 0 ?
+          {
+            key: this.metricKeys[0],
+            isMetric: true
+          } : {
+            key: this.paramKeys[1],
+            isMetric: false
+          }
+      }
+  }
+
+  /**
+   * Find in a list of metrics/params a certain key
+   */
+  static findInList(data, key) {
+    let found = undefined;
+    data.forEach((value) => {
+      if(value.key === key)
+        found = value;
+    });
+    return found;
+  }
+
+  static getKeys(lists) {
+    let keys = {};
+    lists.forEach((list) => 
+      list.forEach((item) => {
+        if(!(item.key in keys))
+          keys[item.key] = true;
+        if(isNaN(parseFloat(item.value)))
+          keys[item.key] = false;
+      }
+    ));
+    return Object.keys(keys).filter(k => keys[k]).sort();
+  }
+
+  // Get the value of the metric/param described by {key, isMetric}, in run i
+  getValue(i, {key, isMetric}) {
+    const value = CompareRunScatter.findInList(
+      (isMetric?this.props.metricLists:this.props.paramLists)[i],
+      key);
```

Comment:
This is formatted a bit weirdly; have spaces around `?` and `:`, and it's ok to keep it on one line if it's less than 100 characters.